### PR TITLE
Next.js: Add experimental SWC support

### LINKED ISSUE #24827: Next.js: Support experimental SWC mode
Add experimental support for SWC in Next.js

### PR DESCRIPTION
**Pull request by valentinpalkovic**
Closes https://github.com/storybookjs/storybook/issues/24827

## What I did

Added experimental SWC mode for Next.js >= 14

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Next.js 12 sandbox

-> Setting `framework.options.builder.useSWC: true` shouldn't have any effect. Instead the user should get a warning in the console, that SWC mode is only supported in Next.js v14 and above. It would be beneficial also to test Next.js 13, since we currently don't have a Next.js 13 sandbox in place. Just check out one of the Next.js sandboxes and change the `next` dependency to the latest v13. The behaviour should be the same as for Next.js v12. SWC is not supported here. Storybook should work as expected in Babel mode.

### Other Next.js sandboxes

-> Set `framework.options.builder.useSWC: true` and test whether all Next.js Stories are working as expected. Personally, I have tested the local font option already.

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

### CODE REVIEW

File: docs/api/main-config-typescript.md [hunk 1]
```diff
@@ -90,6 +90,8 @@ Only available for React Storybook projects. Options to pass to react-docgen-typ
 
 ## `skipBabel`
 
+Deprecated: Will be removed in Storybook 8.0. Use `skipCompiler` instead
+
 Type: `boolean`
 
 Disable parsing of TypeScript files through babel.
```
Comment (jonniebigodes) [inline docs/api/main-config-typescript.md:97]: @valentinpalkovic, when you're able, can you capitalize Babel as we're referring to the tooling and also adjust "Deprecated: Will be removed in Storybook 8.0. Use `skipCompiler` instead"  to "Deprecated: Will be removed in Storybook 8.0. Use `skipCompiler` instead."  as this is a full sentence.

**Review by jonniebigodes — APPROVED** [review]
@valentinpalkovic thanks for taking the time to put together this pull request and include the required changes to the documentation. Appreciate it 🙏 ! Only one small thing for you to look into and this should be good to go from my end. Approving this to unblock you.

File: code/frameworks/nextjs/src/swc/loader.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
+import { getProjectRoot } from '@storybook/core-common';
+import { getVirtualModuleMapping } from '@storybook/core-webpack';
+import type { Options } from '@storybook/types';
+import ReactRefreshWebpackPlugin from '@pmmmwh/react-refresh-webpack-plugin';
+import type { NextConfig } from 'next';
+import path from 'path';
+import type { RuleSetRule } from 'webpack';
+import semver from 'semver';
+import { NextjsSWCNotSupportedError } from 'lib/core-events/src/errors/server-errors';
```
Comment (yannbf) [inline code/frameworks/nextjs/src/swc/loader.ts:9]:
```suggestion
import { NextjsSWCNotSupportedError } from '@storybook/core-events/server-errors';
```

**Review by yannbf — APPROVED** [review]

**Comment by r34son** [comment]
@valentinpalkovic Seems like tracing is removed from original next-swc-loader due to https://github.com/vercel/next.js/pull/65101/files#diff-780a2d4ec404eae0c998ffc7161746a8d26476c1474e0956c1231176ec5fa397

Is it possible to remove patched version of loader?

**Comment by valentinpalkovic** [comment]
@r34son Good catch! Would you like to open a PR?

**Pull request merged**